py-polars/tests/unit/test_df.py

```diff
@@ -2189,11 +2231,15 @@ def test_product() -> None:
             "flt": [-1.0, 12.0, 9.0],
             "bool_0": [True, False, True],
             "bool_1": [True, True, True],
-        }
+        },
+        dtypes={
+            "int": pl.UInt16,
+            "flt": pl.Float32,
+        },
     )
     out = df.product()
     expected = pl.DataFrame({"int": [6], "flt": [-108.0], "bool_0": [0], "bool_1": [1]})
-    assert out.frame_equal(expected)
+    assert_frame_not_equal(out, expected)
```

Comment:
Just out of interest, why not construct `expected` explicitly (like with `.cast`, so it goes down a different path than it would with the `dtypes` constructor) and compare against that? There's an infinite number of arguments that would satisfy `assert_frame_not_equal` 😄

Reply:
Ah; because I probably meant to write this; good catch, will update ;)
```python
    assert_frame_not_equal(out, expected, check_dtype=True)
    assert_frame_equal(out, expected, check_dtype=False)
```